# Tìm Hiểu Về Cronbach's Alpha Từ Dữ Liệu Flash Sale Shopee

---

## 1. Bắt đầu từ một tình huống rất quen với marketer

Giả sử bạn muốn trả lời một câu hỏi rất thực tế:

> **Flash sale có thật sự làm khách hàng thấy gấp, thấy thích thú, hoặc dễ mua ngẫu hứng hơn không?**

Nếu chỉ hỏi một câu, kết quả thường khá mỏng và khó tin cậy. Vì vậy, trong nghiên cứu marketing, người ta thường dùng **một nhóm câu hỏi** để đo cùng một nội dung.

Ví dụ:

- để đo **cảm giác cấp bách**, có thể hỏi nhiều câu về việc sợ hết hàng, sợ bỏ lỡ giá tốt, phải quyết định nhanh,
- để đo **sự thích thú**, có thể hỏi nhiều câu về cảm giác vui vẻ, hạnh phúc, phấn khích.

Khi đã có một nhóm câu hỏi như vậy, ta cần kiểm tra xem chúng có **đủ hợp nhau** để gộp lại thành một thang đo hay không. Đó chính là lúc dùng **Cronbach's Alpha**.

---

## 2. Cronbach's Alpha là gì?

Cronbach's Alpha là một chỉ số giúp trả lời câu hỏi sau:

> **Các câu hỏi trong cùng một nhóm có đang cùng đi theo một ý hay không?**

Hiểu đơn giản:

- **Alpha cao**: các câu hỏi khá ăn khớp với nhau và thường đang cùng phản ánh một nội dung.
- **Alpha thấp**: các câu hỏi chưa hợp nhau lắm, có thể có câu bị lệch ý hoặc đang đo sang phần khác.

Bạn có thể hiểu Cronbach's Alpha như một cách kiểm tra **độ nhất quán nội tại** của thang đo. Nói dễ hiểu hơn, nó cho biết các câu hỏi trong cùng một nhóm có đang "đi cùng nhau" hay không.

Điểm quan trọng cần nhớ:

- Cronbach's Alpha chủ yếu cho biết **mức độ nhất quán giữa các câu hỏi trong cùng một thang đo**.
- Nó **không đủ để kết luận** thang đo là **hợp lệ**.
- Kết quả này luôn phải được hiểu **trên mẫu dữ liệu hiện tại**, không nên xem như một nhãn đúng cho mọi mẫu dữ liệu khác.

Nói cách khác: alpha cao cho thấy trên mẫu đang có, nhóm câu hỏi này phối hợp với nhau khá ổn. Nhưng điều đó không có nghĩa là thang đo chắc chắn hoàn hảo trong mọi trường hợp.

---

## 3. Ví dụ thật từ bộ dữ liệu này

### Ví dụ 1: Thang đo `Urgency`

Trong bộ dữ liệu này, thang đo `Urgency` được đo bằng 5 câu hỏi thật như sau:

- `Urgency_1`: "Tôi cho rằng mình cần phải mua ngay lập tức để sở hữu được sản phẩm này."
- `Urgency_2`: "Tôi cho rằng nếu không mua ngay, sản phẩm sẽ sớm bị bán hết."
- `Urgency_3`: "Tôi cho rằng mình phải quyết định mua nhanh trước khi Flash Sale kết thúc."
- `Urgency_4`: "Tôi cho rằng nếu không mua ngay, tôi có thể bỏ lỡ cơ hội sở hữu sản phẩm với giá ưu đãi."
- `Urgency_5`: "Tôi cho rằng mình sẽ hối tiếc nếu không mua ngay lập tức."

Nếu một người thực sự cảm thấy cấp bách, họ thường sẽ chọn điểm khá cao ở nhiều câu trong nhóm này. Khi đó, các câu hỏi được xem là khá ăn khớp với nhau.

Trong dữ liệu hiện tại, thang đo `Urgency` có Cronbach's Alpha = `0.876`, tức là **độ nhất quán nội tại tốt**.

### Ví dụ 2: Thang đo `Enjoyment`

Thang đo `Enjoyment` trong bộ dữ liệu này gồm 5 câu hỏi:

- `Enjoyment_1`: Flash Sale khiến tôi cảm thấy vui vẻ.
- `Enjoyment_2`: Flash Sale khiến tôi cảm thấy hạnh phúc.
- `Enjoyment_3`: Flash Sale khiến tôi cảm thấy phấn khích.
- `Enjoyment_4`: Flash Sale khiến tôi cảm thấy thích thú.
- `Enjoyment_5`: Flash Sale khiến tôi giảm bớt cảm giác mệt mỏi.

Nếu người trả lời chọn cao ở hầu hết các câu này, điều đó cho thấy nhóm câu hỏi này khá ăn khớp với nhau để đo "cảm xúc tận hưởng".

Trong dữ liệu hiện tại, thang đo `Enjoyment` có Cronbach's Alpha = `0.902`, tức là **độ nhất quán nội tại rất tốt**.

### Ví dụ 3: Một trường hợp alpha thấp (ví dụ minh họa)

Ngoài ví dụ từ dữ liệu thật, ta cũng nên xem một ví dụ giả định về trường hợp alpha thấp để dễ phân biệt.

Đây là **ví dụ giả định**, không phải kết quả thật từ bộ dữ liệu Shopee Flash Sale.

Giả sử ta muốn đo "sự thích thú khi mua flash sale", nhưng lại gom các câu hỏi sau vào cùng một nhóm:

- "Tôi thấy mua flash sale khá vui."
- "Tôi cảm thấy hào hứng khi tham gia flash sale."
- "Flash sale làm trải nghiệm mua sắm thú vị hơn."
- "Tôi thường đọc đánh giá sản phẩm rất kỹ trước khi mua."
- "Tôi quan tâm đến tốc độ giao hàng hơn là giá khuyến mãi."

Trong nhóm này, 3 câu đầu đang nói về **cảm xúc thích thú**, nhưng 2 câu sau lại nghiêng về **thói quen cân nhắc khi mua hàng**.

Vì các câu hỏi không còn cùng đo một ý nữa, Cronbach's Alpha có thể xuống thấp, ví dụ chỉ khoảng `0.55`.

Khi gặp trường hợp như vậy, nhà nghiên cứu nên xem lại:

- các câu hỏi có đang đo cùng một nội dung hay không,
- có câu nào bị lệch ý so với phần còn lại hay không,
- có nên chỉnh lại hoặc loại bớt câu hỏi trước khi phân tích tiếp hay không.

### Ví dụ 4: Câu hỏi đảo chiều nếu không xử lý đúng sẽ làm alpha thấp đi

Đây cũng là **ví dụ giả định**, nhưng rất hay gặp trong nghiên cứu thực tế.

Giả sử bạn có 3 câu cùng đo cảm giác cấp bách:

- "Tôi thấy cần mua ngay khi có flash sale."
- "Tôi sợ hết hàng nếu không mua ngay."
- "Tôi không thấy gấp chút nào khi có flash sale."

Một người thật sự thấy rất cấp bách có thể trả lời như sau:

- câu 1: `5`
- câu 2: `5`
- câu 3: `1`

Về mặt ý nghĩa, câu trả lời này rất hợp lý. Người đó đồng ý mạnh với 2 câu đầu và không đồng ý với câu thứ 3 vì câu thứ 3 đang viết theo hướng ngược lại.

Nhưng nếu giữ nguyên dữ liệu là `5, 5, 1`, phần mềm sẽ dễ hiểu rằng câu thứ 3 đang đi ngược với 2 câu còn lại. Khi đó Cronbach's Alpha có thể bị kéo xuống thấp một cách oan uổng.

Lúc này cần **đảo mã** câu hỏi ngược chiều trước khi tính alpha:

- `1` đổi thành `5`
- `2` đổi thành `4`
- `3` giữ nguyên `3`
- `4` đổi thành `2`
- `5` đổi thành `1`

Sau khi đảo mã, bộ điểm `5, 5, 1` sẽ trở thành `5, 5, 5`. Khi đó phần mềm mới hiểu đúng rằng cả 3 câu đều đang phản ánh cùng một mức độ cấp bách.

Vì vậy, khi thấy alpha thấp, đừng vội kết luận thang đo kém. Trước tiên nên kiểm tra xem có câu hỏi đảo chiều mà chưa được xử lý hay không.

### Ví dụ 5: Nếu mẫu quá nhỏ thì alpha gần như không nói được nhiều

Đây cũng là một tình huống rất quan trọng nhưng dễ bị bỏ qua.

Giả sử bạn có một thang đo gồm 5 câu hỏi, nhưng chỉ có **1 người** trả lời khảo sát.

Người đó trả lời như sau:

- câu 1: `5`
- câu 2: `5`
- câu 3: `4`
- câu 4: `5`
- câu 5: `4`

Nhìn bằng mắt thường, bạn có thể thấy người này trả lời khá nhất quán. Nhưng vấn đề là **Cronbach's Alpha không được tạo ra để đọc từ 1 người trả lời**.

Lý do là alpha cần nhìn sự biến thiên giữa **nhiều người** để xem các câu hỏi có cùng thay đổi với nhau hay không. Nếu chỉ có 1 dòng dữ liệu, gần như không có đủ thông tin để tính phương sai và hiệp phương sai một cách có ý nghĩa.

Nói dễ hiểu hơn:

- `1 người trả lời` thì không thể kết luận thang đo tốt hay không tốt bằng Cronbach's Alpha,
- `rất ít người trả lời` thì dù phần mềm có tính ra một con số, con số đó cũng rất dễ bất ổn,
- mẫu càng đủ lớn và đa dạng, việc diễn giải alpha mới càng đáng tin hơn.

Vì vậy, khi đọc alpha, luôn nên tự hỏi thêm một câu:

> **Con số này được tính trên một mẫu đủ lớn để tin được hay chưa?**

---

## 4. Cách đọc kết quả

| Giá trị alpha | Có thể hiểu nhanh | Nên làm gì? |
|---|---|---|
| >= 0.90 | Rất tốt | Có thể yên tâm dùng cho bước phân tích tiếp theo, nhưng vẫn nên xem các câu có bị quá giống nhau không |
| 0.80-0.89 | Tốt | Có thể tiếp tục phân tích |
| 0.70-0.79 | Chấp nhận được | Thường vẫn dùng được cho các bước tiếp theo |
| 0.60-0.69 | Hơi yếu | Chưa nên yên tâm, cần xem lại câu hỏi, cách mã hóa và câu đảo chiều |
| < 0.60 | Thấp | Không nên dùng ngay, cần kiểm tra lại thang đo trước |

Bạn có thể hiểu theo cách rất thực tế như sau:

- nếu alpha từ `0.70` trở lên, thang đo thường đã đủ ổn để dùng tiếp,
- nếu alpha nằm trong khoảng `0.60` đến dưới `0.70`, nên dừng lại kiểm tra kỹ hơn thay vì dùng ngay,
- nếu alpha dưới `0.60`, nên xem lại thang đo trước khi chạy các phân tích tiếp theo.

Khi alpha thấp, nên kiểm tra lần lượt:

- các câu hỏi có thật sự cùng đo một nội dung hay không,
- có câu nào viết ngược chiều mà chưa đảo mã hay không,
- có câu nào khó hiểu hoặc dễ bị người trả lời hiểu theo cách khác hay không,
- mẫu dữ liệu có quá nhỏ để việc diễn giải alpha trở nên thiếu tin cậy hay không.

Ví dụ:

- Nếu `Urgency = 0.876`, có thể nói thang đo này có độ nhất quán nội tại tốt và có thể tiếp tục sử dụng.
- Nếu `Enjoyment = 0.902`, có thể nói thang đo này có độ nhất quán nội tại rất tốt.
- Nếu một thang đo chỉ đạt `0.58`, chưa nên dùng ngay. Trước tiên hãy kiểm tra lại nội dung câu hỏi và cách mã hóa.

Lưu ý thêm: nếu alpha quá cao, ví dụ trên `0.95`, cũng nên xem lại vì có thể các câu hỏi đang quá giống nhau và lặp ý nhiều.

---

## 5. Khi nào nên dùng Cronbach's Alpha?

Cronbach's Alpha phù hợp khi:

- một nội dung được đo bằng **nhiều câu hỏi Likert**,
- các câu hỏi trong nhóm đều đang hướng tới **cùng một ý**,
- các câu hỏi được mã hóa theo **cùng một chiều**.

Cronbach's Alpha không phù hợp khi:

- chỉ có một câu hỏi,
- đó là thông tin mô tả mẫu như giới tính, tuổi, nghề nghiệp,
- các câu hỏi trong cùng nhóm thực ra đang đo nhiều nội dung khác nhau,
- số quan sát quá ít để phản ánh được sự biến thiên giữa người trả lời.

Nếu có câu hỏi mang ý nghĩa ngược với các câu còn lại, cần đảo mã trước khi tính alpha.

---

## 6. Bảng giải thích thuật ngữ

| Thuật ngữ | Nghĩa đơn giản | Ví dụ trong bộ dữ liệu này |
|---|---|---|
| Khái niệm | Nội dung mà nghiên cứu muốn đo | `Urgency`, `Enjoyment`, `Impulse` |
| Biến quan sát | Một câu hỏi cụ thể trong bảng khảo sát | `Urgency_1`, `Urgency_2` |
| Thang đo | Một nhóm câu hỏi cùng đo một khái niệm | 5 câu hỏi của `Urgency` |
| Thang Likert | Câu hỏi có trả lời theo mức độ đồng ý | `1 = Rất không đồng ý` đến `5 = Rất đồng ý` |
| Độ nhất quán nội tại | Mức độ các câu hỏi trong cùng nhóm có đi cùng nhau hay không | Các câu hỏi của `Enjoyment` cùng phản ánh cảm xúc tận hưởng |
| Tính hợp lệ | Thang đo có thật sự đo đúng điều cần đo không | `Urgency` có thật sự đo cảm giác cấp bách không |
| Cronbach's Alpha | Chỉ số kiểm tra các câu hỏi trong cùng nhóm có ăn khớp với nhau không | Alpha của `Enjoyment` hoặc `Urgency` |
| Mã hóa ngược chiều | Câu hỏi có ý nghĩa ngược với các câu còn lại và cần đảo mã trước khi phân tích | Nếu có câu kiểu "Tôi không thấy áp lực khi có flash sale" |

---

## 7. Bộ dữ liệu đang dùng trong notebook này

Phần tính Cronbach's Alpha trong notebook này chạy trực tiếp trên file:

- `data/Survey result_Flash sale.sav`

File `data/Survey data collection.xlsx` chỉ đóng vai trò nguồn đối chiếu khi cần giải thích câu hỏi gốc hoặc kiểm tra lại mapping biến.

Trong file `.sav`, mỗi dòng là một người trả lời khảo sát trong mẫu đã làm sạch và mỗi cột là một biến đã được mã hóa.

- Các câu hỏi thang Likert được mã hóa từ `1` đến `5`
- Với các câu hỏi chọn nhiều đáp án, mỗi lựa chọn được tách thành một cột riêng với giá trị `1 = Có chọn`, `0 = Không chọn`

Ghi chú quan trọng:

- `data/Survey result_Flash sale.sav` không phải toàn bộ file Excel gốc, mà là mẫu đã được làm sạch và lọc còn `267` quan sát.
- Ở nhóm câu hỏi chọn nhiều đáp án, nên ưu tiên tên cột và codebook thay vì chỉ nhìn metadata của SPSS, vì có ít nhất một nhãn bị gán trùng (`Shopee_Reason_Variety`).

---

## 8. Các thang đo có trong bộ dữ liệu

| Thang đo | Các cột trong `.sav` | Ý nghĩa ngắn gọn |
|---|---|---|
| `Time_Limit` | `Limit_Time_1`, `Limit_Time_2`, `Time_Limit_3`, `Time_Limit_4`, `Time_Limit_5` | Nhận thức về giới hạn thời gian |
| `Qty_Scarcity` | `Qty_Scarcity_1` đến `Qty_Scarcity_5` | Nhận thức về giới hạn số lượng sản phẩm |
| `Voucher_Scarcity` | `Voucher_Scarcity_1` đến `Voucher_Scarcity_5` | Nhận thức về giới hạn số lượng voucher |
| `Visual` | `Visual_1` đến `Visual_5` | Nhận thức về khả năng hiển thị |
| `Eco_Benefit` | `Eco_Benefit_1` đến `Eco_Benefit_5` | Nhận thức về lợi ích kinh tế |
| `Social_Proof` | `Social_Proof_1` đến `Social_Proof_5` | Nhận thức về bằng chứng xã hội |
| `Urgency` | `Urgency_1` đến `Urgency_5` | Nhận thức về sự cấp bách |
| `Enjoyment` | `Enjoyment_1` đến `Enjoyment_5` | Cảm xúc tận hưởng |
| `Impulse` | `Impulse_1` đến `Impulse_5` | Hành vi mua sắm ngẫu hứng |

Nếu cần xem **từng cột trong file `.sav` map với câu hỏi gốc nào**, hãy xem file codebook: `data/SAV_CODEBOOK.md`.


In [6]:
import pyreadstat

# Đọc bộ dữ liệu đã được làm sạch và mã hóa từ SPSS.
df, meta = pyreadstat.read_sav('data/Survey result_Flash sale.sav')
print(f'Số dòng: {df.shape[0]}, Số cột: {df.shape[1]}')
df.head()


Số dòng: 267, Số cột: 79


,STT,Shopee_Exp,Shopee_Freq,Shopee_Monthly_Spend,Shopee_Reason_Price,Shopee_Reason_Variety,Shopee_Reason_Promotion,Shopee_Reason_Interface,Shopee_Reason_Payment,Shopee_Reason_Delivery,...,Impulse_3,Impulse_4,Impulse_5,Gender,Age,Education,Occupation,Income,Marital_Status,Residence
0,1.0,2.0,2.0,4.0,1.0,1.0,0.0,0.0,1.0,0.0,...,4.0,5.0,3.0,1.0,2.0,4.0,4.0,3.0,3.0,1.0
1,2.0,5.0,4.0,4.0,1.0,1.0,1.0,0.0,0.0,0.0,...,4.0,2.0,2.0,2.0,2.0,5.0,4.0,4.0,3.0,1.0
2,3.0,5.0,2.0,1.0,1.0,1.0,1.0,0.0,0.0,0.0,...,3.0,3.0,3.0,2.0,2.0,5.0,4.0,2.0,3.0,1.0
3,4.0,5.0,3.0,4.0,1.0,0.0,1.0,1.0,0.0,0.0,...,3.0,2.0,3.0,2.0,2.0,5.0,4.0,2.0,3.0,1.0
4,5.0,5.0,2.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0,...,2.0,2.0,2.0,1.0,2.0,5.0,4.0,2.0,3.0,1.0


In [7]:
import pandas as pd


def prepare_items(items_df):
    return items_df.apply(pd.to_numeric, errors='coerce').dropna()


def cronbach_alpha(items_df):
    k = items_df.shape[1]
    if k < 2:
        raise ValueError("Cronbach's Alpha requires at least two items.")

    item_variances = items_df.var(axis=0, ddof=1).sum()
    total_variance = items_df.sum(axis=1).var(ddof=1)
    if pd.isna(total_variance) or total_variance == 0:
        raise ValueError("Total score variance must be greater than 0 to compute Cronbach's Alpha.")

    return (k / (k - 1)) * (1 - item_variances / total_variance)


def interpret_alpha(alpha):
    if alpha >= 0.90:
        return 'Rất tốt'
    if alpha >= 0.80:
        return 'Tốt'
    if alpha >= 0.70:
        return 'Chấp nhận được'
    if alpha >= 0.60:
        return 'Hơi yếu, cần xem lại'
    return 'Yếu, cần điều chỉnh'


scales = {
    'Eco_Benefit': [f'Eco_Benefit_{i}' for i in range(1, 6)],
    'Enjoyment': [f'Enjoyment_{i}' for i in range(1, 6)],
    'Impulse': [f'Impulse_{i}' for i in range(1, 6)],
    'Qty_Scarcity': [f'Qty_Scarcity_{i}' for i in range(1, 6)],
    'Social_Proof': [f'Social_Proof_{i}' for i in range(1, 6)],
    'Urgency': [f'Urgency_{i}' for i in range(1, 6)],
    'Visual': [f'Visual_{i}' for i in range(1, 6)],
    'Voucher_Scarcity': [f'Voucher_Scarcity_{i}' for i in range(1, 6)],
    'Time_Limit': ['Limit_Time_1', 'Limit_Time_2', 'Time_Limit_3', 'Time_Limit_4', 'Time_Limit_5'],
}

results = []

for scale_name, columns in scales.items():
    scale_df = prepare_items(df[columns])
    alpha = cronbach_alpha(scale_df)
    results.append(
        {
            'Thang đo': scale_name,
            'Số biến': len(columns),
            'Số quan sát hợp lệ': len(scale_df),
            'Cronbach Alpha': round(alpha, 3),
            'Diễn giải': interpret_alpha(alpha),
        }
    )

results_df = pd.DataFrame(results).sort_values('Cronbach Alpha', ascending=False).reset_index(drop=True)
results_df


,Thang đo,Số biến,Số quan sát hợp lệ,Cronbach Alpha,Diễn giải
0,Impulse,5,267,0.910,Rất tốt
1,Voucher_Scarcity,5,267,0.908,Rất tốt
2,Enjoyment,5,267,0.902,Rất tốt
3,Social_Proof,5,267,0.879,Tốt
4,Urgency,5,267,0.876,Tốt
5,Visual,5,267,0.863,Tốt
6,Eco_Benefit,5,267,0.860,Tốt
7,Qty_Scarcity,5,267,0.842,Tốt
8,Time_Limit,5,267,0.839,Tốt


## 9. Mẫu Câu Viết Vào Bài

Dựa trên bộ dữ liệu hiện tại, có thể viết như sau:

> Cronbach's Alpha được sử dụng để kiểm tra độ nhất quán nội tại của các thang đo trong nghiên cứu. Trên mẫu dữ liệu hiện tại gồm 267 quan sát, kết quả cho thấy giá trị alpha dao động từ 0.839 đến 0.910, cho thấy tất cả các thang đo đều có độ nhất quán nội tại từ tốt đến rất tốt. Vì vậy, các thang đo được giữ lại để sử dụng trong các phân tích tiếp theo.

Một câu ngắn gọn, an toàn và đúng là:

> **Trên mẫu dữ liệu hiện tại, các thang đo đều có độ nhất quán nội tại tốt và phù hợp để sử dụng cho các phân tích tiếp theo.**

Nếu cần giải thích tên cột và câu hỏi gốc chi tiết hơn, hãy mở file `data/SAV_CODEBOOK.md`.
